File: Bchain.js
'use strict';
const shim = require('fabric-shim');
const util = require('util');

class Client{
  constructor(Name,ClientID,clientType,SkAccountNumber,Currency,Asset){
      this.Name=Name;
      this.ClientID=ClientID;
      this.clientType=clientType;
      this.SkAccountNumber=SkAccountNumber;
      this.Currency=Currency;
      this.Asset=Asset;
      this.Status="active";985B32676EB9A4B9ED0B5D03907065D788E265387C7563C7BFA436DCEC5CEFB0
  }
}

class Asset{
	constructor(AssetName,AssetID,AssetAmount){
      this.AssetName=AssetName;
      this.AssetID=AssetID;
      this.AssetAmount=AssetAmount;
      }	
}

class Transaction{
	constructor(TsID,CF,CT,Asset,TimeStamp){
      this.TsID=TsID;
      this.CF=CF;
      this.CT=CT;
      this.Asset=Asset;
      this.TimeStamp=TimeStamp
  }
}

let Chaincode = class {
  async Init(stub) {
    console.info('=========== Instantiated fabcar chaincode ===========');
    return shim.success();
  }

  async Invoke(stub) {
    let ret = stub.getFunctionAndParameters();
    console.info(ret);

    let method = this[ret.fcn];
    if (!method) {
      console.error('no function of name:' + ret.fcn + ' found');
      throw new Error('Received unknown function ' + ret.fcn + ' invocation');
    }
    try {
      let payload = await method(stub, ret.params);
      return shim.success(payload);
    } catch (err) {
      console.log(err);
      return shim.error(err);
    }
  }

  async queryClientInfo(stub, args) {
    if (args.length != 1) {
      throw new Error('Incorrect number of arguments. Expecting CarNumber ex: CAR01');
    }
    let carNumber = args[0];

    let carAsBytes = await stub.getState(carNumber); //get the car from chaincode state
    if (!carAsBytes || carAsBytes.toString().length <= 0) {
      throw new Error(carNumber + ' does not exist: ');
    }
    console.log(carAsBytes.toString());
    return carAsBytes;
  }

  async initLedger(stub, args) {
    console.info('============= START : Initialize Ledger ===========');
    let cars = [];
	  let asset1=[];
    asset1.push(new Asset("ATVI","A1",2500));
    asset1.push(new Asset("BABA","A2",2500));
    asset1.push(new Asset("JD","A3",2500));
    asset1.push(new Asset("AMD","A4",2500));
	  asset1.push(new Asset("INTEL","A5",2500));
    cars.push(new Client("CITI ADMINISTRATION","AD000001","Admin","USD","348912452",asset1));

    let asset2=[];
    asset2.push(new Asset("ATVI","A1",1500));
    asset2.push(new Asset("BABA","A2",1500));
    asset2.push(new Asset("JD","A3",1500));
    asset2.push(new Asset("AMD","A4",1500));
	  asset2.push(new Asset("INTEL","A5",1500));
    cars.push(new Client("THE GREENWALL FOUNDATION","20180001","Regular","USD","348912975",asset2));
	
	  let asset3=[];
    asset3.push(new Asset("ATVI","A1",500));
    asset3.push(new Asset("BABA","A2",500));
    asset3.push(new Asset("JD","A3",500));
    asset3.push(new Asset("AMD","A4",500));
	  asset3.push(new Asset("INTEL","A5",500));
    cars.push(new Client("SOLAR CAPITAL LTD","20180002","Regular","USD","348912325",asset3));

	  let asset4=[];
    asset4.push(new Asset("ATVI","A1",1500));
    asset4.push(new Asset("BABA","A2",2500));
    asset4.push(new Asset("JD","A3",3500));
    cars.push(new Client("PFPC-DFA FUNDS-IRISH","20180003","Regular","USD","348912345",asset4));

    let asset5=[];
    asset5.push(new Asset("AMD","A4",3500));
    asset5.push(new Asset("INTEL","A5",5500));
    cars.push(new Client("ORBIS GROUP","20180004","Regular","USD","348912790",asset5));
    for (let i = 0; i < cars.length; i++) {
      await stub.putState('CLIENT' + i, Buffer.from(JSON.stringify(cars[i])));
      console.info('Added <--> ', cars[i]);
    }
    console.info('============= END : Initialize Ledger ===========');
  }

  async createClient(stub, args) {
    console.info('============= START : Create Car ===========');
    if (args.length != 7) {
      throw new Error('Incorrect number of arguments. Expecting 7');
    }

    var client = new Client(args[1],args[2],args[3],args[4],args[5],args[6]);

    await stub.putState(args[0], Buffer.from(JSON.stringify(client)));
    console.info('============= END : Create Car ===========');
  }

  async queryAllAsset(stub, args) {

    let startKey = 'CLIENT0';
    let endKey = 'CLIENT999';

    let iterator = await stub.getStateByRange(startKey, endKey);

    let allResults = [];
    while (true) {
      let res = await iterator.next();

      if (res.value && res.value.value.toString()) {
        let jsonRes = {};
        console.log(res.value.value.toString('utf8'));

        jsonRes.Key = res.value.key;
        try {
          jsonRes.Record = JSON.parse(res.value.value.toString('utf8'));
        } catch (err) {
          console.log(err);
          jsonRes.Record = res.value.value.toStringAssetID('utf8');
        }
        allResults.push(jsonRes);
      }
      if (res.done) {
        console.log('end of data');
        await iterator.close();
        console.info(allResults);
        return Buffer.from(JSON.stringify(allResults));
      }
    }
  }

  async doInnerTransaction(stub, args) {
    //CF client key//CT //Asset ID/Amount
    if (args.length != 4) {
      throw new Error('Incorrect number of arguments:'+args.length+'. Expecting 4:[BankIDFrom,BankIDTo,AssetName,AssetAmount]');
    }
    let startKey = 'TRANC1';
	  let endKey = 'TRANC999';
	  let iterator = await stub.getStateByRange(startKey, endKey);
    let count=1;
	  while (true) {
     let res = await iterator.next();
     if (res.value && res.value.value.toString()) {
        count=parseInt(count)+1;
     }
     else{
       break;
     }
    }
    let transaction_id='TRANC'+count;

    let bankFAsBytes = await stub.getState(args[0]);
    let bankTAsBytes = await stub.getState(args[1]);
    let bankF = JSON.parse(bankFAsBytes);
    let bankT = JSON.parse(bankTAsBytes);
    let asset_amount_from=0;
    var asset_from;
    let asset_exist_in_to=false;
    for(var asset in bankF.Asset){
     if(bankF.Asset[asset].AssetID==args[2] && parseFloat(bankF.Asset[asset].AssetAmount)>= parseFloat(args[3])){
      asset_amount_from=bankF.Asset[asset].AssetAmount;
      asset_from=bankF.Asset[asset];
      bankF.Asset[asset].AssetAmount=parseFloat(bankF.Asset[asset].AssetAmount) - parseFloat(args[3]);
     }
    }
    if(asset_amount_from==0 || parseFloat(asset_amount_from) < parseFloat(args[3])){
     throw new Error('Incorrect number of arguments. Expecting Correct Asset'+asset_amount_from);
    }
    else{
     for(var asset_to in bankT.Asset){
      if(bankT.Asset[asset_to].AssetID==args[2]){
        bankT.Asset[asset_to].AssetAmount=parseFloat(bankT.Asset[asset_to].AssetAmount) + parseFloat(args[3]);
        asset_exist_in_to=true;
      }
     }
     if(!asset_exist_in_to){
      bankT.Asset.push(new Asset(asset_from.AssetName,asset_from.AssetID,args[3]));
     }
     // registre transaction
     await stub.putState(args[0], Buffer.from(JSON.stringify(bankF)));
     await stub.putState(args[1], Buffer.from(JSON.stringify(bankT)));
     await stub.putState(transaction_id, Buffer.from(JSON.stringify( new Transaction(transaction_id,args[0],args[1],new Asset(asset_from.AssetName,asset_from.AssetID,args[3]),new Date()))));
    }
    console.info('============= END : changeCarowner ===========');
  }

  async changeClientStatus(stub, args) {
    console.info('============= START : changeCarowner ===========');
    if (args.length != 2) {
      throw new Error('Incorrect number of arguments. Expecting 2');
    }

    let carAsBytes = await stub.getState(args[0]);
    let car = JSON.parse(carAsBytes);
    car.Status = args[1];

    await stub.putState(args[0], Buffer.from(JSON.stringify(car)));
    console.info('============= END : changeCarowner ===========');
  }

  async queryTransactions(stub, args) {
    console.info('============= START : queryTransactions ===========');
    if (args.length != 1) {
      throw new Error('Incorrect number of arguments. Expecting ID like :CLIENT0');
    }
    let findall=false;
    if(args[0]=="ALL"){
      findall=true;
    }
	  let startKey = 'TRANC1';
	  let endKey = 'TRANC999';
	  let iterator = await stub.getStateByRange(startKey, endKey);
	  let allResults = [];
	  while (true) {
		  let res = await iterator.next();
		  if (res.value && res.value.value.toString()) {
			let jsonRes = {};
			console.log(res.value.value.toString('utf8'));

			jsonRes.TransactionID = res.value.key;
			try {
			  jsonRes.Record = JSON.parse(res.value.value.toString('utf8'));
			} catch (err) {
			  console.log(err);
			  jsonRes.Record = res.value.value.toString('utf8');
			}
			if(findall || jsonRes.Record.CF==args[0] || jsonRes.Record.CT==args[0]){
                 allResults.push(jsonRes);
			  }
		  }
		  if (res.done) {
			console.log('end of data');
			await iterator.close();
			console.info(allResults);
			return Buffer.from(JSON.stringify(allResults));
		  }
	  }
  }
};

shim.start(new Chaincode());

File: Bchain.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"strconv"
	"time"

	"github.com/hyperledger/fabric/core/chaincode/shim"
	sc "github.com/hyperledger/fabric/protos/peer"
)

type SmartContract struct {
}

type Assets struct {
	AssetName   string  `json:"assetname"`
	AssetID     string  `json:"assetid"`
	AssetAmount float64 `json:"assetamount"`
}

type Clients struct {
	Name            string   `json:"name"`
	ClientID        string   `json:"clientid"`
	ClientType      string   `json:"clienttype"`
	SkAccountNumber string   `jsoasset_ton:"skaccountnumber"`
	Currency        string   `json:"currency"`
	Asset           []Assets `json:"asset"`
	Status          string   `json:"status"`
}

type Transactions struct {
	TsID      string `json:"tsid"`
	CF        string `json:"cf"`
	CT        string `json:"ct"`
	Asset     Assets `json:"asset"`
	TimeStamp string `json:"timestamp"`
}

func (s *SmartContract) Invoke(APIstub shim.ChaincodeStubInterface) sc.Response {

	// Retrieve the requested Smart Contract function and arguments
	function, args := APIstub.GetFunctionAndParameters()
	// Route to the appropriate handler function to interact with the ledger appropriately
	if function == "createClient" {
		return s.createClient(APIstub, args)
	} else if function == "initLedger" {
		return s.initLedger(APIstub)
	} else if function == "doInnerTransaction" {
		return s.doInnerTransaction(APIstub, args)
	} else if function == "changeClientStatus" {
		return s.changeClientStatus(APIstub, args)
	} else if function == "queryTransactions" {
		return s.queryTransactions(APIstub, args)
	} else if function == "queryClientInfo" {
		return s.queryClientInfo(APIstub, args)
	}
	return shim.Error("Invalid Smart Contract function name.")
}

func (s *SmartContract) queryClientInfo(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
	if len(args) != 1 {
		return shim.Error("Incorrect number of arguments. Expecting 1")
	}
	carAsBytes, _ := APIstub.GetState(args[0])
	return shim.Success(carAsBytes)
}

func (s *SmartContract) initLedger(APIstub shim.ChaincodeStubInterface) sc.Response {
	asset1 := []Assets{
		Assets{AssetName: "ATVI", AssetID: "A1", AssetAmount: 2500.00},
		Assets{AssetName: "BABA", AssetID: "A2", AssetAmount: 2500.00},
		Assets{AssetName: "JD", AssetID: "A3", AssetAmount: 2500.00},
		Assets{AssetName: "AMD", AssetID: "A4", AssetAmount: 2500.00},
		Assets{AssetName: "INTEL", AssetID: "A5", AssetAmount: 2500.00},
	}

	asset2 := []Assets{
		Assets{AssetName: "ATVI", AssetID: "A1", AssetAmount: 1500.00},
		Assets{AssetName: "BABA", AssetID: "A2", AssetAmount: 1500.00},
		Assets{AssetName: "JD", AssetID: "A3", AssetAmount: 1500.00},
		Assets{AssetName: "AMD", AssetID: "A4", AssetAmount: 1500.00},
		Assets{AssetName: "INTEL", AssetID: "A5", AssetAmount: 1500.00},
	}

	asset3 := []Assets{
		Assets{AssetName: "ATVI", AssetID: "A1", AssetAmount: 500.00},
		Assets{AssetName: "BABA", AssetID: "A2", AssetAmount: 500.00},
		Assets{AssetName: "JD", Asasset_tosetID: "A3", AssetAmount: 500.00},
		Assets{AssetName: "AMD", AssetID: "A4", AssetAmount: 500.00},
		Assets{AssetName: "INTEL", AssetID: "A5", AssetAmount: 500.00},
	}

	asset4 := []Assets{
		Assets{AssetName: "ATVI", AssetID: "A1", AssetAmount: 1500.00},
		Assets{AssetName: "BABA", AssetID: "A2", AssetAmount: 2500.00},
		Assets{AssetName: "JD", AssetID: "A3", AssetAmount: 3500.00},
	}

	asset5 := []Assets{
		Assets{AssetName: "AMD", AssetID: "A4", AssetAmount: 3500.00},
		Assets{AssetName: "INTEL", AssetID: "A5", AssetAmount: 5500.00},
	}

	var clients = []Clients{
		Clients{Name: "CITI ADMINISTRATION", ClientID: "AD000001", ClientType: "Admin", SkAccountNumber: "348912452", Currency: "USD", Asset: asset1, Status: "active"},
		Clients{Name: "THE GREENWALL FOUNDATION", ClientID: "20180001", ClientType: "Regular", SkAccountNumber: "348912975", Currency: "USD", Asset: asset2, Status: "active"},
		Clients{Name: "SOLAR CAPITAL LTD", ClientID: "20180002", ClientType: "Regular", SkAccountNumber: "348912325", Currency: "USD", Asset: asset3, Status: "active"},
		Clients{Name: "PFPC-DFA FUNDS-IRISH", ClientID: "20180003", ClientType: "Regular", SkAccountNumber: "348912345", Currency: "USD", Asset: asset4, Status: "active"},
		Clients{Name: "ORBIS GROUP", ClientID: "20180004", ClientType: "Regular", SkAccountNumber: "348912790", Currency: "USD", Asset: asset5, Status: "active"},
	}

	i := 0
	for i < len(clients) {
		fmt.Println("i is ", i)
		clientAsBytes, _ := json.Marshal(clients[i])
		APIstub.PutState("CLIENT"+strconv.Itoa(i), clientAsBytes)
		fmt.Println("Added", clients[i])
		i = i + 1
	}

	return shim.Success(nil)
}

/*func (s *SmartContract) createClient(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

  if (args.length != 7) {
    throw new Error('Incorrect number of arguments. Expecting 7');
  }

  var client = new Client(args[1],args[2],args[3],args[4],args[5],args[6]);

  await stub.putState(args[0], Buffer.from(JSON.stringify(client)));
  console.info('============= END : Create Car ===========');
}*/

func (s *SmartContract) queryAllAsset(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	startKey := "CLIENT0"
	endKey := "CLIENT999"

	resultsIterator, err := APIstub.GetStateByRange(startKey, endKey)
	if err != nil {
		return shim.Error(err.Error())
	}
	defer resultsIterator.Close()

	var buffer bytes.Buffer
	buffer.WriteString("[")

	bArrayMemberAlreadyWritten := false
	for resultsIterator.HasNext() {
		queryResponse, err := resultsIterator.Next()
		if err != nil {
			return shim.Error(err.Error())
		}
		// Add a comma before array members, suppress it for the first array member
		if bArrayMemberAlreadyWritten == true {
			buffer.WriteString(",")
		}
		buffer.WriteString("{\"Key\":")
		buffer.WriteString("\"")
		buffer.WriteString(queryResponse.Key)
		buffer.WriteString("\"")

		buffer.WriteString(", \"Record\":")
		// Record is a JSON object, so we write as-is
		buffer.WriteString(string(queryResponse.Value))
		buffer.WriteString("}")
		bArrayMemberAlreadyWritten = true
	}
	buffer.WriteString("]")

	fmt.Printf("- queryAllAsset:\n%s\n", buffer.String())

	return shim.Success(buffer.Bytes())
}

func (s *SmartContract) doInnerTransaction(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
	//CF client key//CT //Asset ID/Amount
	if len(args) != 4 {
		//return shim.Error('Incorrect number of arguments:'+len(args)+'. Expecting 4:[BankIDFrom,BankIDTo,AssetName,AssetAmount]')
	}
	startcarKey := "TRANC1"
	endKey := "TRANC999"
	resultsIterator, err := APIstub.GetStateByRange(startKey, endKey)
	if err != nil {
		return shim.Error(err.Error())
	}
	defer resultsIterator.Close()

	count := 1
	for resultsIterator.HasNext() {
		queryResponse, err := resultsIterator.Next()
		if err != nil {
			return shim.Error(err.Error())
		}
		if queryResponse.Value != nil {
			count = count + 1
		} else {
			break
		}
	}
	transaction_id := "TRANC" + count

	bankFAsBytes, _ := APIstub.GetState(args[0])
	bankTAsBytes, _ := APIstub.GetState(args[1])

	bankF := Clients{}
	bankT := Clients{}
	json.Unmarshal(bankFAsBytes, &bankF)
	json.Unmarshal(bankTAsBytes, &bankT)
	var asset_amount_from float64
	asset_from := Assets{}
	var asset_exist_in_to = false
	amount, _ := strconv.ParseFloat(args[3], 64)
	for i, asset := range bankF.Asset {
		if asset.AssetID == args[2] && asset.AssetAmount >= amount {
			asset_amount_from = asset.AssetAmount
			asset_from = asset
			bankF.Asset[i].AssetAmount = asset.AssetAmount - amount
		}
	}
	if asset_amount_from == 0 || asset_amount_from < amount {
		return shim.Error("Incorrect number of arguments. Expecting Correct Asset" + asset_amount_from)
	} else {
		for i, asset_to := range bankT.Asset {
			if asset_to.AssetID == args[2] {
				asset_from = asset_to
				bankT.Asset[i].AssetAmount = asset_to.AssetAmount + amount
				asset_exist_in_to = true
			}
		}
	}
	if !asset_exist_in_to {
		bankT.Asset = append(bankT.Asset, Assets{AssetName: asset_from.AssetName, AssetID: asset_from.AssetID, AssetAmount: amount})
	}

	// registre transaction
	bankFAsBytes_after, _ := json.Marshal(bankF)
	bankTAsBytes_after, _ := json.Marshal(bankT)
	transactionAsBytes, _ = json.Marshal(Transactions{TsID: transaction_id, CF: args[0], CT: args[1], Asset: Assets{AssetName: asset_from.AssetName, AssetID: asset_from.AssetID, AssetAmount: amount}, TimeStamp: string(time.Now().Unix())})

	APIstub.PutState(args[0], bankFAsBytes_after)
	APIstub.PutState(args[1], bankTAsBytes_after)
	APIstub.PutState(transaction_id, transactionAsBytes)
}

func (s *SmartContract) queryTransactions(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
	if len(args) != 1 {
		return shim.Error("Incorrect number of arguments. Expecting ID like :CLIENT0")
	}
	findall := false
	if args[0] == "ALL" {
		findall = true
	}
	startKey := "TRANC1"
	endKey := "TRANC999"
	resultsIterator, err := APIstub.GetStateByRange(startKey, endKey)
	if err != nil {
		return shim.Error(err.Error())
	}
	defer resultsIterator.Close()
	var allResults []Transactions = make([]Transactions, 5)
	for resultsIterator.HasNext() {
		queryResponse, err := resultsIterator.Next()
		objres := Transactions{}
		if err != nil {
			return shim.Error(err.Error())
		}
		json.Unmarshal(string(queryResponse.Value), &objres)
		if findall || objres.CF == args[0] || objres.CT == args[0] {
			allResults = append(allResults, objres)
		}
	}
	transactionsAsBytes, _ = json.Marshal(allResults)
	return shim.Success(transactionsAsBytes)
}

func main() {
	// Create a new Smart Contract
	err := shim.Start(new(SmartContract))
	if err != nil {
		fmt.Printf("Error creating new Smart Contract: %s", err)
	}
}
